gui/src/latching_alarm.h

```diff
@@ -0,0 +1,78 @@
+#ifndef LATCHING_ALARM_H_
+#define LATCHING_ALARM_H_
+
+#include "chrono.h"
+#include "network_protocol.pb.h"
+
+#include <optional>
+
+// An alarm with a latching audio signal that can be silenced.
+// Has two signals: visual and audio.
+// * visual signal is non-latching: it is active when the condition is active.
+// * audio signal is latching, i.e. stays active until suppressed even if
+//   condition is not active anymore.
+// Audio can be suppressed for up to 2 minutes via Acknowledge().
+// Suppression applies to the current occurrence, i.e. it expires either after
+// 2 minutes or when the condition becomes inactive, whichever is earlier.
+// The terminology is taken from
+// https://github.com/RespiraWorks/VentilatorSoftware/wiki/Alarm-Subsystem.
+class LatchingAlarm {
+private:
+  virtual bool IsActive(SteadyInstant now, const ControllerStatus &status) = 0;
+
+public:
+  virtual ~LatchingAlarm() = default;
+
+  // Updates the state of the alarm and its signals according to current
+  // sensor readings.
+  void Update(SteadyInstant now, const ControllerStatus &status) {
+    currently_active_ = IsActive(now, status);
+    switch (audio_state_) {
+    case AudioState::INACTIVE:
+      audio_state_ =
+          currently_active_ ? AudioState::BEEPING : AudioState::INACTIVE;
+      break;
+    case AudioState::BEEPING:
+      audio_state_ = AudioState::BEEPING;
+      break;
+    case AudioState::SILENCED:
+      if (now > silenced_until_) {
+        silenced_until_ = std::nullopt;
+        audio_state_ =
+            currently_active_ ? AudioState::BEEPING : AudioState::INACTIVE;
+      } else {
+        audio_state_ = currently_active_ ? AudioState::SILENCED
+                                         : (silenced_until_ = std::nullopt,
+                                            AudioState::INACTIVE);
+      }
+      break;
+    }
```

Comment:
Ugh, this `operator,` business is clever.

Reply:
Yeah this is possibly the first time I've used it in my life, but it seemed to look more readable than plain conditionals. I rewrote this quite a few times and this is the most readable form I came up with.